utils/cmd/cmd.c

```diff
@@ -0,0 +1,721 @@
+				goto finish;
+			}
+			written += res;
+		}
+	}
+
+	if (output == fd_pcm)
+		bluealsa_dbus_pcm_ctrl_send_drain(fd_pcm_ctrl, &err);
+	else
+		/* sleep arbitrary 300ms to allow stdout to drain */
+		usleep(300000);
+
+finish:
+	close(fd_pcm);
+	close(fd_pcm_ctrl);
+
+	return EXIT_SUCCESS;
+}
+
+static DBusHandlerResult dbus_signal_handler(DBusConnection *conn, DBusMessage *message, void *data) {
+	(void)conn;
+	(void)data;
+
+	if (dbus_message_get_type(message) != DBUS_MESSAGE_TYPE_SIGNAL)
+		return DBUS_HANDLER_RESULT_NOT_YET_HANDLED;
+
+	const char *interface = dbus_message_get_interface(message);
+	const char *signal = dbus_message_get_member(message);
+
+	DBusMessageIter iter;
+
+	if (strcmp(interface, BLUEALSA_INTERFACE_MANAGER) == 0) {
+
+		if (strcmp(signal, "PCMAdded") == 0)
+			if (dbus_message_iter_init(message, &iter) &&
+					dbus_message_iter_get_arg_type(&iter) == DBUS_TYPE_OBJECT_PATH) {
+				const char *path;
+				dbus_message_iter_get_basic(&iter, &path);
+				printf("PCMAdded %s\n", path);
+				return DBUS_HANDLER_RESULT_HANDLED;
+			}
+
+		if (strcmp(signal, "PCMRemoved") == 0)
+			if (dbus_message_iter_init(message, &iter) &&
+					dbus_message_iter_get_arg_type(&iter) == DBUS_TYPE_OBJECT_PATH) {
+				const char *path;
+				dbus_message_iter_get_basic(&iter, &path);
+				printf("PCMRemoved %s\n", path);
+				return DBUS_HANDLER_RESULT_HANDLED;
+			}
+	}
+
+	return DBUS_HANDLER_RESULT_NOT_YET_HANDLED;
+}
+
+static bool monitor_loop_on = true;
+static void monitor_loop_stop(int sig) {
+	/* Call to this handler restores the default action, so on the
+	 * second call the program will be forcefully terminated. */
+
+	struct sigaction sigact = { .sa_handler = SIG_DFL };
+	sigaction(sig, &sigact, NULL);
+
+	monitor_loop_on = false;
+}
+
+static int monitor(int argc, char *argv[]) {
+	if (argc != 1) {
+		print_error_usage("Invalid arguments.");
+		return EXIT_FAILURE;
+	}
+
+	bluealsa_dbus_connection_signal_match_add(&dbus_ctx,
+			options.service, NULL, BLUEALSA_INTERFACE_MANAGER, "PCMAdded", NULL);
+	bluealsa_dbus_connection_signal_match_add(&dbus_ctx,
+			options.service, NULL, BLUEALSA_INTERFACE_MANAGER, "PCMRemoved", NULL);
+
+	if (!dbus_connection_add_filter(dbus_ctx.conn, dbus_signal_handler, NULL, NULL)) {
+		print_error("Couldn't add D-Bus filter");
+		return EXIT_FAILURE;
+	}
+
+	struct sigaction sigact = { .sa_handler = monitor_loop_stop };
+	sigaction(SIGTERM, &sigact, NULL);
+	sigaction(SIGINT, &sigact, NULL);
+
+	while (monitor_loop_on) {
+
+		struct pollfd pfds[10];
+		nfds_t pfds_len = ARRAYSIZE(pfds);
+
+		if (!bluealsa_dbus_connection_poll_fds(&dbus_ctx, pfds, &pfds_len)) {
+			print_error("Couldn't get D-Bus connection file descriptors");
+			return EXIT_FAILURE;
+		}
+
+		if (poll(pfds, pfds_len, -1) == -1 &&
+				errno == EINTR)
+			continue;
+
+		if (bluealsa_dbus_connection_poll_dispatch(&dbus_ctx, pfds, pfds_len))
+			while (dbus_connection_dispatch(dbus_ctx.conn) == DBUS_DISPATCH_DATA_REMAINS)
```

Comment:
Making the change you suggest here, I get 100% CPU busy-loop. I've never done any D-Bus API coding before, so only have a vague understanding of what is expected here. Will look into it again soon.

Reply:
My bad, I forgot that this function does not block :/ I will make some changes to the dbus-client.c code so it will be easier to use it in this PR. Also, I will change this function (`bluealsa_dbus_connection_dispatch()`) to allow blocking mode.

Reply:
My bad part 2 :D Use `dbus_connection_read_write_dispatch()` https://dbus.freedesktop.org/doc/api/html/group__DBusConnection.html#ga580d8766c23fe5f49418bc7d87b67dc6

But now the question is why I have crated `bluealsa_dbus_connection_dispatch()` in the first place....

Reply:
see commit 1d92201